### PR TITLE
Unique formUrl Creation

### PR DESCRIPTION
**Pull request by McCambley**
# Overview

1) Add `formUrl` field to `Investor` Model
2) Use [NanoId](https://github.com/ai/nanoid/#mongoose) to create random ID
3) Create `async` function to verify uniqueness of ID before `Investor` creation.
   - If a user with the ID already exists, a new ID is created.

**Notes:** [This handy calculator](https://zelark.github.io/nano-id-cc/) helps determine the likelihood that we'll have colliding ID's. With `8` character long ID's, the chance that we'll have colliding ID's is extremely small. Still, it could be smaller. Increase the string length to `10` and we buy ourselves a lot more time. As we add more character beyond `8`, however, what we gain in uniqueness, we lose in the simplicity of the Investor's URL. Fortunately, the `createUniqueUrl` function ensures that even _if_ we do have colliding ID's, the duplicate will be scrapped and a new one will be created in its place (over and over until a unique one is created). 

[Here's a loom demo to see it all in action
](https://www.loom.com/share/da8c698cdf5249d89940b1b6c4006dfb)
# I follow best practices 

### Calculator Screenshot 
Screen Shot 2021-12-02 at 7 08 22 PM

- [ ] My code is covered by unit tests (working on it)
- [x] This PR is less than [300 lines of code](https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests)
- [x] My code [smells good](https://en.wikipedia.org/wiki/Code_smell) 
- [x] I have tested my PR locally and proven it works. 
- [x] There is [no copy-paste](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

### CODE REVIEW

**Comment by vgalilei** [comment]
@McCambley You manual tests in video - worth to write as unit test :-) this is a really good example where you create new algorithm.

File: backend/controllers/investors.js [hunk 1]
```diff
@@ -1,7 +1,27 @@
...
+  const newUrl = nanoid(8);
+  // check if user with url already exists
+  return Investor.findOne({ formUrl: newUrl })
+    .then((investor) => {
+      if (!investor) {
+        // if one doesn't, return the formUrl
+        return newUrl;
+      }
+      // if one does exist, create a new url and check again,
+      return createUniqueUrl();
```
Comment (vgalilei) [inline backend/controllers/investors.js:16]: I can suggest one improvement: 
One another cool option is to pass charactersCount to this function and increment it on each recursion call. 
Hence 
* you can start from 2 and have very tiny URLs
* your recursion produce unique Id with less than < 10 iterations for sure

**Review by vgalilei — APPROVED** [review]
LGTM after adding unittests

File: backend/controllers/investors.js (hunk 1 above)
Comment (McCambley) [reply backend/controllers/investors.js:16]: Ohhh wow I like that idea. Something like this?
```javascript
let characterCount = 2; // could be any number

function createUniqueUrl() {
  // create new formUrl with 8 characters
  const newUrl = nanoid(characterCount);
  // check if user with url already exists
  return Investor.findOne({ formUrl: newUrl })
    .then((investor) => {
      if (!investor) {
        // if one doesn't, return the formUrl
        return newUrl;
      }
      // if one does exist, create a new url and check again,
      characterCount += 1;
      return createUniqueUrl();
    })
    .catch((error) => {
      console.error(error);
    });
}
```

**Comment by McCambley** [comment]
> LGTM after adding unittests

I'm going to write test in a new PR. In writing the new one I was able to find a fix that will allow for the other test to become more robust. I'll merge this then write more tests tomorrow.

File: backend/controllers/investors.js (hunk 1 above)
Comment (vgalilei) [reply backend/controllers/investors.js:19]: @McCambley  i mean `createUniqueUrl(characterCount)` and then  `createUniqueUrl(characterCount+1)`

**Pull request merged**